{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "def read_fasta_file(filename):\n",
    "    \"\"\"\n",
    "    Reads the given FASTA file f and returns a dictionary of sequences.\n",
    "\n",
    "    Lines starting with ';' in the FASTA file are ignored.\n",
    "    \"\"\"\n",
    "    sequences_lines = {}\n",
    "    current_sequence_lines = None\n",
    "    with open(filename) as fp:\n",
    "        for line in fp:\n",
    "            line = line.strip()\n",
    "            if line.startswith(';') or not line:\n",
    "                continue\n",
    "            if line.startswith('>'):\n",
    "                sequence_name = line.lstrip('>')\n",
    "                current_sequence_lines = []\n",
    "                sequences_lines[sequence_name] = current_sequence_lines\n",
    "            else:\n",
    "                if current_sequence_lines is not None:\n",
    "                    current_sequence_lines.append(line)\n",
    "    sequences = {}\n",
    "    for name, lines in sequences_lines.items():\n",
    "        sequences[name] = ''.join(lines)\n",
    "    return sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "a=read_fasta_file(\"true-ann1.fa\")\n",
    "\n",
    "def map_truestate_to_3statemarkov(strn):\n",
    "    out=[]\n",
    "    for i in strn:\n",
    "        if i == 'C':\n",
    "            out.append(0)\n",
    "        elif i == 'N':\n",
    "            out.append(1)\n",
    "        elif i == 'R':\n",
    "            out.append(2)\n",
    "        else:\n",
    "            raise Exception(\"Input should be a either N, C or R\")\n",
    "    return out\n",
    "\n",
    "def map_truestate_to_7statemarkov(strn):\n",
    "    out=[None]*len(strn)\n",
    "    for i in range(len(strn)):\n",
    "        if strn[i] == 'C':\n",
    "            if out[i-1]==0:\n",
    "                out[i]=(1)\n",
    "            elif out[i-1]==1:\n",
    "                out[i]=(2)\n",
    "            else:\n",
    "                out[i]=(0)\n",
    "        elif strn[i] == 'N':\n",
    "            out[i]=(3)\n",
    "        elif strn[i] == 'R':\n",
    "            if out[i-1]==4:\n",
    "                out[i]=(5)\n",
    "            elif out[i-1]==5:\n",
    "                out[i]=(6)\n",
    "            else:\n",
    "                out[i]=(4)\n",
    "        else:\n",
    "            raise Exception(\"Input should be a either N, C or R\")\n",
    "    return out\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def translate_observations_to_indices(obs):\n",
    "    mapping = {'a': 0, 'c': 1, 'g': 2, 't': 3}\n",
    "    return [mapping[symbol.lower()] for symbol in obs]\n",
    "\n",
    "def translate_indices_to_observations(indices):\n",
    "    mapping = ['a', 'c', 'g', 't']\n",
    "    return ''.join(mapping[idx] for idx in indices)\n",
    "\n",
    "class hmm:\n",
    "    def __init__(self, init_probs, trans_probs, emission_probs):\n",
    "        self.init_probs = init_probs\n",
    "        self.trans_probs = trans_probs\n",
    "        self.emission_probs = emission_probs\n",
    "\n",
    "\n",
    "\n",
    "init_probs_3_state = np.array(\n",
    "    [0.00, 1.00, 0.00]\n",
    ")\n",
    "\n",
    "trans_probs_3_state = np.array([\n",
    "    [0.90, 0.10, 0.00],\n",
    "    [0.05, 0.90, 0.05],\n",
    "    [0.00, 0.10, 0.90],\n",
    "])\n",
    "\n",
    "emission_probs_3_state = np.array([\n",
    "    #   A     C     G     T\n",
    "    [0.40, 0.15, 0.20, 0.25],\n",
    "    [0.25, 0.25, 0.25, 0.25],\n",
    "    [0.20, 0.40, 0.30, 0.10],\n",
    "])\n",
    "\n",
    "hmm_3_state = hmm(init_probs_3_state,\n",
    "                  trans_probs_3_state,\n",
    "                  emission_probs_3_state)\n",
    "\n",
    "init_probs_7_state = np.array(\n",
    "    [0.00, 0.00, 0.00, 1.00, 0.00, 0.00, 0.00]\n",
    ")\n",
    "\n",
    "\n",
    "\n",
    "trans_probs_7_state = np.array([\n",
    "    [0.90, 0.00, 0.00, 0.10, 0.00, 0.00, 0.00],\n",
    "    [0.00, 1.00, 0.00, 0.00, 0.00, 0.00, 0.00],\n",
    "    [0.00, 0.00, 1.00, 0.00, 0.00, 0.00, 0.00],\n",
    "    [0.05, 0.00, 0.00, 0.90, 0.05, 0.00, 0.00],\n",
    "    [0.00, 0.00, 0.00, 0.10, 0.90, 0.00, 0.00],\n",
    "    [0.00, 0.00, 0.00, 0.00, 0.00, 1.00, 0.00],\n",
    "    [0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 1.00],    \n",
    "])\n",
    "\n",
    "emission_probs_7_state = np.array([\n",
    "    #   A     C     G     T\n",
    "    [0.30, 0.25, 0.25, 0.20],\n",
    "    [0.20, 0.35, 0.15, 0.30],\n",
    "    [0.40, 0.15, 0.20, 0.25],\n",
    "    [0.25, 0.25, 0.25, 0.25],\n",
    "    [0.20, 0.40, 0.30, 0.10],\n",
    "    [0.30, 0.20, 0.30, 0.20],\n",
    "    [0.15, 0.30, 0.20, 0.35],\n",
    "])\n",
    "\n",
    "hmm_7_state = hmm(init_probs_7_state,\n",
    "                  trans_probs_7_state,\n",
    "                  emission_probs_7_state)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import math\n",
    "def log(x):\n",
    "    if x == 0:\n",
    "        return float(\"-inf\")\n",
    "    else:\n",
    "        return math.log(x)\n",
    "\n",
    "#Fill out the blanks to get the Viterbi algorithm:\n",
    "\n",
    "def viterbi(obs, hmm):\n",
    "    X = translate_observations_to_indices(obs)\n",
    "    N = len(X)\n",
    "    K = len(hmm.init_probs)\n",
    "    V = np.zeros((K,N))\n",
    "\n",
    "    init_probs=np.log(hmm.init_probs)\n",
    "    trans_probs=np.log(hmm.trans_probs)\n",
    "    emission_probs=np.log(hmm.emission_probs)\n",
    "\n",
    "    for i in range(K):\n",
    "        V[i][0]=init_probs[i]\n",
    "    for i in range(1, N):\n",
    "        if i%100000==0:\n",
    "            print(i) \n",
    "        for n in range(K):\n",
    "            E = emission_probs[n][X[i]]\n",
    "            T = trans_probs[:,n]+V[:,i-1]\n",
    "            V[n][i]= E + np.max(T)\n",
    "    return V\n",
    "\n",
    "\n",
    "def backtrack(V, hmm):\n",
    "    assert len(V) == len(hmm.init_probs)\n",
    "\n",
    "    init_probs = np.log(hmm.init_probs)\n",
    "    trans_probs = np.log(hmm.trans_probs)\n",
    "    emission_probs = np.log(hmm.emission_probs)\n",
    "\n",
    "    N = len(V[0])\n",
    "    i = N - 1\n",
    "    k = np.argmax(V[:,N-1])\n",
    "    o = []\n",
    "\n",
    "    while i >= 0:\n",
    "        o.append(k)\n",
    "        k = np.argmax(V[:,i-1] + trans_probs[:,k] )\n",
    "        i-=1\n",
    "        if i % 100000 == 0: print(i)\n",
    "    return o[::-1]\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_accuracy(true_ann, pred_ann):\n",
    "    if len(true_ann) != len(pred_ann):\n",
    "        return 0.0\n",
    "    return sum(1 if true_ann[i] == pred_ann[i] else 0 \n",
    "               for i in range(len(true_ann))) / len(true_ann)\n",
    "\n",
    "true_ann = map_truestate_to_7statemarkov(read_fasta_file(\"true-ann1.fa\")[\"true-ann1\"])\n",
    "\n",
    "obs = read_fasta_file(\"genome1.fa\")\n",
    "pred_ann = backtrack(viterbi(obs[\"genome1\"], hmm_7_state),hmm_7_state)\n",
    "\n",
    "compute_accuracy(true_ann, pred_ann)\n",
    "\n",
    "#question 1 = 0.2546272728794061"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "true_ann = map_truestate_to_7statemarkov(read_fasta_file(\"true-ann2.fa\")[\"true-ann2\"])\n",
    "\n",
    "obs = read_fasta_file(\"genome2.fa\")\n",
    "pred_ann = backtrack(viterbi(obs[\"genome2\"], hmm_7_state),hmm_7_state)\n",
    "\n",
    "compute_accuracy(true_ann, pred_ann)\n",
    "\n",
    "#question 2 = 0.2181276382159499"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "true_ann = map_truestate_to_3statemarkov(read_fasta_file(\"true-ann1.fa\")[\"true-ann1\"])\n",
    "\n",
    "obs = read_fasta_file(\"genome1.fa\")\n",
    "pred_ann = backtrack(viterbi(obs[\"genome1\"], hmm_3_state),hmm_3_state)\n",
    "\n",
    "compute_accuracy(true_ann, pred_ann)\n",
    "\n",
    "#question 3 = 0.31873349812490653"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "true_ann = map_truestate_to_3statemarkov(read_fasta_file(\"true-ann2.fa\")[\"true-ann2\"])\n",
    "\n",
    "obs = read_fasta_file(\"genome2.fa\")\n",
    "pred_ann = backtrack(viterbi(obs[\"genome2\"], hmm_3_state),hmm_3_state)\n",
    "\n",
    "compute_accuracy(true_ann, pred_ann)\n",
    "\n",
    "#question 4 = 0.35088368223162264"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Attempt at training"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#We train the init_probs by setting them to their true initial state, this is fairly straight-forward as we know the init-state from the true-ann1.fa file\n",
    "\n",
    "true_str = read_fasta_file(\"true-ann1.fa\")[\"true-ann1\"]\n",
    "\n",
    "def train_init_state(true_str, states):\n",
    "    if states == 3:\n",
    "        init_state=map_truestate_to_3statemarkov(true_str[0])\n",
    "        I = np.array(\n",
    "            [0.00, 1.00, 0.00]\n",
    "            )\n",
    "    elif states == 7:\n",
    "        init_state=map_truestate_to_7statemarkov(true_str[0])\n",
    "        I = np.array(\n",
    "            [0.00, 0.00, 0.00, 1.00, 0.00, 0.00, 0.00]\n",
    "            )\n",
    "\n",
    "    return I\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#to train the trans_probs we need to make a count of all the transitions.\n",
    "\n",
    "true_str = read_fasta_file(\"true-ann1.fa\")[\"true-ann1\"]\n",
    "\n",
    "def train_trans_probs(true_str, states):\n",
    "    if states == 3:\n",
    "        true_str = map_truestate_to_3statemarkov(true_str)\n",
    "    elif states == 7:\n",
    "        true_str = map_truestate_to_7statemarkov(true_str)\n",
    "    else:\n",
    "        raise Exception(\"State number not supported\")\n",
    "    \n",
    "    T = np.zeros((states,states))\n",
    "    for i in range(1,len(true_str)):\n",
    "        T[true_str[i-1]][true_str[i]]+=1\n",
    "    for i in range(states):\n",
    "        counts = 0\n",
    "        for n in range(states):\n",
    "            counts+=T[i][n]\n",
    "        for n in range(states):\n",
    "            T[i][n]=T[i][n]/counts\n",
    "\n",
    "    return(T)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "true_str = read_fasta_file(\"true-ann1.fa\")[\"true-ann1\"]\n",
    "nuc_str = read_fasta_file(\"genome1.fa\")[\"genome1\"]\n",
    "\n",
    "def train_emission_probs(nuc_str,true_str, states):\n",
    "    if states == 3:\n",
    "        true_lst = map_truestate_to_3statemarkov(true_str)\n",
    "    elif states == 7:\n",
    "        true_lst = map_truestate_to_7statemarkov(true_str)\n",
    "    else:\n",
    "        raise Exception(\"State number not supported\")\n",
    "    \n",
    "    nuc_str = translate_observations_to_indices(nuc_str)\n",
    "\n",
    "    T = np.zeros((states,4))\n",
    "\n",
    "    for i in range(1,len(true_lst)):\n",
    "        T[true_lst[i]][nuc_str[i]]+=1\n",
    "    for i in range(states):\n",
    "        counts = 0\n",
    "        for n in range(4):\n",
    "            counts+=T[i][n]\n",
    "        for n in range(4):\n",
    "            T[i][n]=T[i][n]/counts\n",
    "\n",
    "    return(T)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#and put it all together\n",
    "\n",
    "def trained_hmm(true_str, nuc_str, states):\n",
    "    out=hmm(train_init_state(true_str, states),\n",
    "    train_trans_probs(true_str,states),\n",
    "    train_emission_probs(nuc_str,true_str, states))\n",
    "    return out\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "true_str1 = read_fasta_file(\"true-ann1.fa\")[\"true-ann1\"]\n",
    "nuc_str1 = read_fasta_file(\"genome1.fa\")[\"genome1\"]\n",
    "\n",
    "hmm_3_state1=trained_hmm(true_str1,nuc_str1,3)\n",
    "\n",
    "hmm_7_state1=trained_hmm(true_str1,nuc_str1,7)\n",
    "\n",
    "\n",
    "true_str2 = read_fasta_file(\"true-ann2.fa\")[\"true-ann2\"]\n",
    "nuc_str2 = read_fasta_file(\"genome2.fa\")[\"genome2\"]\n",
    "\n",
    "hmm_3_state2=trained_hmm(true_str2,nuc_str2,3)\n",
    "\n",
    "hmm_7_state2=trained_hmm(true_str2,nuc_str2,7)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "true_ann = map_truestate_to_7statemarkov(read_fasta_file(\"true-ann2.fa\")[\"true-ann2\"])\n",
    "\n",
    "obs = read_fasta_file(\"genome2.fa\")\n",
    "pred_ann = backtrack(viterbi(obs[\"genome2\"], hmm_7_state1),hmm_7_state1)\n",
    "\n",
    "print(len(true_ann),len(pred_ann))\n",
    "\n",
    "compute_accuracy(true_ann, pred_ann)\n",
    "\n",
    "#question 5 = 0.7800618136681913"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "true_ann = map_truestate_to_3statemarkov(read_fasta_file(\"true-ann2.fa\")[\"true-ann2\"])\n",
    "\n",
    "obs = read_fasta_file(\"genome2.fa\")\n",
    "pred_ann = backtrack(viterbi(obs[\"genome2\"], hmm_3_state1),hmm_3_state1)\n",
    "\n",
    "print(len(true_ann),len(pred_ann))\n",
    "\n",
    "compute_accuracy(true_ann, pred_ann)\n",
    "\n",
    "#question 6 = 0.57368917266"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "true_ann = map_truestate_to_7statemarkov(read_fasta_file(\"true-ann1.fa\")[\"true-ann1\"])\n",
    "\n",
    "obs = read_fasta_file(\"genome1.fa\")\n",
    "pred_ann = backtrack(viterbi(obs[\"genome1\"], hmm_7_state2),hmm_7_state2)\n",
    "\n",
    "print(len(true_ann),len(pred_ann))\n",
    "\n",
    "compute_accuracy(true_ann, pred_ann)\n",
    "\n",
    "#question 7 = 0.7608863116288184"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "true_ann = map_truestate_to_3statemarkov(read_fasta_file(\"true-ann1.fa\")[\"true-ann1\"])\n",
    "\n",
    "obs = read_fasta_file(\"genome1.fa\")\n",
    "pred_ann = backtrack(viterbi(obs[\"genome1\"], hmm_3_state2),hmm_3_state2)\n",
    "\n",
    "print(len(true_ann),len(pred_ann))\n",
    "\n",
    "compute_accuracy(true_ann, pred_ann)\n",
    "\n",
    "#question 8 = 0.5909796857227841"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "6b4edfe8f5b63aad8d5fd22163b1b34a2a814ee0f50c40673c64e6089f508cae"
  },
  "kernelspec": {
   "display_name": "Python 3.8.8 64-bit ('base': conda)",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
